jar.go

```diff
@@ -295,6 +299,101 @@ func (j *Jar) cookies(u *url.URL, now time.Time) (cookies []*http.Cookie) {
 	return cookies
 }
 
+// AllCookies returns all cookies for the given URL, or a subpath of the given URL.
+// The returned cookies will have Domain, Expires, HttpOnly, Name,
+// Secure, Path, and Value filled out.
+//
+// It returns an empty slice if the URL's scheme is not HTTP or HTTPS.
+func (j *Jar) AllCookies(u *url.URL) (cookies []*http.Cookie) {
+	return j.allCookies(u, time.Now())
+}
+
+// allCookies is like AllCookies but takes the current time as a parameter.
+func (j *Jar) allCookies(u *url.URL, now time.Time) (cookies []*http.Cookie) {
+	if u.Scheme != "http" && u.Scheme != "https" {
+		return cookies
+	}
+	host, err := canonicalHost(u.Host)
+	if err != nil {
+		return cookies
+	}
+	key := jarKey(host, j.psList)
+
+	j.mu.Lock()
+	defer j.mu.Unlock()
+
+	submap := j.entries[key]
+	if submap == nil {
+		return cookies
+	}
+
+	path := u.Path
+	if path == "" {
+		path = "/"
+	}
+	if len(path) > 1 && strings.HasSuffix(path, "/") {
+		path = strings.TrimSuffix(path, "/")
+	}
+
+	var selected []entry
+	for id, e := range submap {
+		if !e.Expires.After(now) {
+			// Save some space by deleting the value when the cookie
+			// expires. We can't delete the cookie itself because then
+			// we wouldn't know that the cookie had expired when
+			// we merge with another cookie jar.
+			if e.Value != "" {
```

Comment:
I'd be tempted to comment the function that this will also delete expired cookies, it's a bit side-effecty, we should tell people in comments